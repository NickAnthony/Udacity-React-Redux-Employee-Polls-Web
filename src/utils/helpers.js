import { useLocation, useNavigate, useParams } from "react-router-dom";

// Copied directly from https://reactrouter.com/en/main/start/faq
export function withRouter(Component) {
  function ComponentWithRouterProp(props) {
    let location = useLocation();
    let navigate = useNavigate();
    let params = useParams();
    return <Component {...props} router={{ location, navigate, params }} />;
  }

  return ComponentWithRouterProp;
}

export function formatTimestamp(timestamp) {
  const date = new Intl.DateTimeFormat("en-US", {
    year: "numeric",
    month: "numeric",
    day: "numeric",
  }).format(timestamp);
  const time = new Intl.DateTimeFormat("en-US", {
    hour: "numeric",
    minute: "2-digit",
  }).format(timestamp);
  return `${time} | ${date}`;
}
