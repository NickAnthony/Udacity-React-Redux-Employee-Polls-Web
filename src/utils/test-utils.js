import { configureStore } from "@reduxjs/toolkit";
import reducers from "../reducers";
import middleware from "../middleware";
import { users, questions } from "../utils/_DATA";
import CreateQuestion from "../components/CreateQuestion";
import CreateUser from "../components/CreateUser";
import Dashboard from "../components/Dashboard";
import Leaderboard from "../components/Leaderboard";
import Login from "../components/Login";
import Question from "../components/Question";
import UserProfile from "../components/UserProfile";
import { Provider } from "react-redux";
import { MemoryRouter, Route, Routes } from "react-router-dom";

export function createTestStore(initialState) {
  const store = configureStore({
    reducer: reducers,
    middleware: middleware,
    preloadedState: initialState,
  });
  return store;
}

export function createTestStoreWithInitialData() {
  return createTestStore({
    users,
    questions,
    authedUser: "nickanthony",
  });
}

export function createTestAppWithStore(initialRoute) {
  return (
    <Provider store={createTestStoreWithInitialData()}>
      <MemoryRouter initialEntries={[initialRoute]}>
        <Routes>
          <Route path="/poll/:id" element={<Question />} />
        </Routes>
      </MemoryRouter>
    </Provider>
  );
}
