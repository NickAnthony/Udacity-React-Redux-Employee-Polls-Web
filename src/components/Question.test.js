import { render, fireEvent } from "@testing-library/react";
import { createTestAppWithStoreAndRouter } from "../utils/test-utils";

describe("Question", () => {
  it("will display the vote counts by clicking either of the vote buttons.", () => {
    var component = render(
      createTestAppWithStoreAndRouter("/poll/8xf0y6ziyjabvozdd253nd")
    );
    // Verify we have 2 vote options and no answer options
    expect(component.queryByTestId("answered-option")).not.toBeInTheDocument();
    expect(component.getAllByTestId("vote-option-button").length).toEqual(2);

    // Get one of the two vote options
    const voteButton = component.getAllByTestId("vote-option-button")[0];
    fireEvent.click(voteButton);

    // Verify we have 2 vote options and no answer options
    expect(
      component.queryByTestId("vote-option-button")
    ).not.toBeInTheDocument();
    expect(component.getAllByTestId("answered-option").length).toEqual(2);
    expect(component.getByTestId("user-answer")).toBeInTheDocument();
  });

  it("will match the question snapshot.", () => {
    var component = render(
      createTestAppWithStoreAndRouter("/poll/8xf0y6ziyjabvozdd253nd")
    );
    expect(component).toMatchSnapshot();
  });
});
