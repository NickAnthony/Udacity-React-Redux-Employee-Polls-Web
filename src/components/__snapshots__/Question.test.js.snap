// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question will match the question snapshot. 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container question-layout-container positive-top-margin"
      >
        <h3>
          Poll by 
          sarahedo
        </h3>
        <div
          class="profile-avatar"
          style="background-image: url(DefaultAvatar.png); height: 120px; width: 120px;"
        />
        <h2>
          Would You Rather...
        </h2>
        <div
          class="container-row-start"
        >
          <div
            class="container-column question-option-container green-bold-outline"
            data-testid="answered-option"
          >
            <span
              class="question-option"
            >
              Build our new application with Javascript
            </span>
            <span
              class="question-answer-percentage"
            >
              Voting: 
              100
              % (
              2
               out of 
              2
              )
            </span>
            <span
              class="green-background"
            >
              "Your vote"
            </span>
          </div>
          <div
            class="container-column question-option-container "
            data-testid="answered-option"
          >
            <span
              class="question-option"
            >
              Build our new application with Typescript
            </span>
            <span
              class="question-answer-percentage"
            >
              Voting: 
              0
              % (
              0
               out of 
              2
              )
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container question-layout-container positive-top-margin"
    >
      <h3>
        Poll by 
        sarahedo
      </h3>
      <div
        class="profile-avatar"
        style="background-image: url(DefaultAvatar.png); height: 120px; width: 120px;"
      />
      <h2>
        Would You Rather...
      </h2>
      <div
        class="container-row-start"
      >
        <div
          class="container-column question-option-container green-bold-outline"
          data-testid="answered-option"
        >
          <span
            class="question-option"
          >
            Build our new application with Javascript
          </span>
          <span
            class="question-answer-percentage"
          >
            Voting: 
            100
            % (
            2
             out of 
            2
            )
          </span>
          <span
            class="green-background"
          >
            "Your vote"
          </span>
        </div>
        <div
          class="container-column question-option-container "
          data-testid="answered-option"
        >
          <span
            class="question-option"
          >
            Build our new application with Typescript
          </span>
          <span
            class="question-answer-percentage"
          >
            Voting: 
            0
            % (
            0
             out of 
            2
            )
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
